### PR TITLE
Legend 2: Create legend components

### PR DESCRIPTION
**Pull request by yvonnesjy**

### CODE REVIEW

File: src/js/models/maps/assets/MapAsset.js [hunk 1]
```diff
@@ -315,64 +308,61 @@ define([
...
-            this.set(
-              "colorPalette",
-              new AssetColorPalette(assetConfig.colorPalette),
-            );
-          }
+        let parsedAssetConfig;
+        if (!assetConfig || typeof assetConfig !== "object") {
+          parsedAssetConfig = {};
+        } else {
+          parsedAssetConfig = JSON.parse(JSON.stringify(assetConfig));
```
Comment (ianguerin) [inline src/js/models/maps/assets/MapAsset.js:321]: JSON.parse throws, I think a try catch may still be necessary at least around this line

File: src/js/templates/maps/legend/layer-legend.html [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<div class="layer-legend__layer-name"><%= name %></div>
+<div class="<%= classNames.palette %>">palette</div>
```
Comment (ianguerin) [inline src/js/templates/maps/legend/layer-legend.html:2]: Is this just a placeholder for work in progress?

File: src/js/views/maps/legend/LegendContainerView.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        });
+      },
+
+      toggleExpanded() {
+        this.expanded = !this.expanded;
+        if (this.expanded) {
+          this.$el.addClass("expanded");
+        } else {
+          this.$el.removeClass("expanded");
+        }
```
Comment (ianguerin) [inline src/js/views/maps/legend/LegendContainerView.js:75]: Could simplify this to `this.$el.toggleClass("expanded");`

File: src/js/views/maps/legend/LegendContainerView.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      },
+
+      updateLegend() {
+        this.$(`.${CLASS_NAMES.content}`).empty();
```
Comment (ianguerin) [inline src/js/views/maps/legend/LegendContainerView.js:79]: Cache this jQuery element to a variable for use on line 88?

**Review by ianguerin — APPROVED** [review]

File: src/js/templates/maps/legend/layer-legend.html (hunk 2 above)
Comment (yvonnesjy) [reply src/js/templates/maps/legend/layer-legend.html:2]: Yes! The implementation is in https://github.com/NCEAS/metacatui/pull/2445 and https://github.com/NCEAS/metacatui/pull/2446. Sorry that wasn't very clear

File: src/js/models/maps/assets/MapAsset.js (hunk 1 above)
Comment (yvonnesjy) [reply src/js/models/maps/assets/MapAsset.js:321]: looking at the method again- there's no altering assetConfig, so no reason to parse(stringify()) it. changing this to a direct assignment.

Comment (yvonnesjy) [reply src/js/models/maps/assets/MapAsset.js]: Done

File: src/js/views/maps/legend/LegendContainerView.js (hunk 3 above)
Comment (yvonnesjy) [reply src/js/views/maps/legend/LegendContainerView.js]: OH! Done

File: src/js/views/maps/legend/LegendContainerView.js (hunk 4 above)
Comment (yvonnesjy) [reply src/js/views/maps/legend/LegendContainerView.js]: Nice! Done

**Review by robyngit — CHANGES_REQUESTED** [review]

File: src/js/models/maps/assets/MapAsset.js [hunk 5]
```diff
@@ -328,13 +328,18 @@ define([
...
-            );
-          }
+        // Set the color palette
+        if (assetConfig.colorPalette) {
+          this.set(
+            "colorPalette",
+            new AssetColorPalette(
+              _.extend(
+                assetConfig.colorPalette,
+                _.pick(assetConfig, "label"),
```
[the remark this answers is not in the document]
Comment (yvonnesjy) [reply src/js/models/maps/assets/MapAsset.js]: Ah yea I saw that palette's label is currently not used so I assumed it's safe to override. But I just noticed that it's mentioned in ColorPaletteConfig so other portals might have it configured. I updated it to use the layer label only if a label is not set in the color palette config.

File: src/js/views/maps/legend/LegendContainerView.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        );
+
+        this.updateLegend();
+        this.model.get("allLayers")?.forEach((layer) => {
+          this.stopListening(layer, "change:visible");
+          this.listenTo(layer, "change:visible", this.updateLegend);
+        });
+      },
+
+      toggleExpanded() {
```
Comment (robyngit) [inline src/js/views/maps/legend/LegendContainerView.js:66]:
```suggestion
      /**
       * Toggles the expanded state of the legend container.
       */
      toggleExpanded() {
```

File: src/js/models/maps/Map.js [hunk 7]
```diff
@@ -143,7 +143,7 @@ define([
       /**
        * Coordinates that describe a camera position for Cesium. Requires at
        * least a longitude and latitude.
-       * @typedef {object} MapConfig#CameraPosition
+       * @typedef {object} CameraPosition
```
[the remark this answers is not in the document]
Comment (yvonnesjy) [reply src/js/models/maps/Map.js:146]: yea. the @property tag doesn't like `#` in the type name, so I was looking for the simplest way to make it stop complaining. that said, I'm realizing there're other places that still use MapConfig#CameraPosition and might break the doc now. I'm okay with just revert these typedef changes to not blow up the scope of this PR. In the future, we might want to consider using @namespace instead https://jsdoc.app/tags-namespace

File: src/js/views/maps/legend/LegendContainerView.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        this.$el.toggleClass("expanded");
+      },
+
+      updateLegend() {
```
Comment (robyngit) [inline src/js/views/maps/legend/LegendContainerView.js:70]:
```suggestion
      /**
       * Updates the legend with the current color palettes.
       */
      updateLegend() {
```

File: src/js/models/maps/Map.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (robyngit) [reply src/js/models/maps/Map.js]: Great idea! https://github.com/NCEAS/metacatui/issues/2450

**Review by robyngit — APPROVED** [review]

**Pull request merged**